packages/cw0/src/parse_reply.rs

```diff
@@ -0,0 +1,214 @@
+    let str_field = parse_protobuf_length_prefixed(data)?;
+    if str_field.is_empty() {
+        return Err(ParseReplyError::ParseFailure(
+            "failed to decode Protobuf message: string field: message too short".to_owned(),
+        ));
+    }
+    Ok(String::from_utf8(str_field)?)
+}
+
+fn parse_protobuf_bytes(data: &mut Vec<u8>) -> Result<Option<Binary>, ParseReplyError> {
+    let bytes_field = parse_protobuf_length_prefixed(data)?;
+    if bytes_field.is_empty() {
+        Ok(None)
+    } else {
+        Ok(Some(Binary(bytes_field)))
+    }
+}
+
+pub fn parse_reply_instantiate_data(
+    msg: Reply,
+) -> Result<MsgInstantiateContractResponse, ParseReplyError> {
+    let data = msg
+        .result
+        .into_result()
+        .map_err(ParseReplyError::SubMsgFailure)?
+        .data
+        .ok_or_else(|| ParseReplyError::ParseFailure("Missing reply data".to_owned()))?;
+    // Manual protobuf decoding
+    let mut data = data.0;
+    // Parse contract addr
+    let contract_addr = parse_protobuf_string(&mut data)?;
+
+    // Parse (optional) data
+    let data = parse_protobuf_bytes(&mut data)?;
+
+    Ok(MsgInstantiateContractResponse {
+        contract_address: contract_addr,
+        data,
+    })
+}
+
+pub fn parse_reply_execute_data(msg: Reply) -> Result<MsgExecuteContractResponse, ParseReplyError> {
+    let data = msg
+        .result
+        .into_result()
+        .map_err(ParseReplyError::SubMsgFailure)?
+        .data
+        .ok_or_else(|| ParseReplyError::ParseFailure("Missing reply data".to_owned()))?;
+    // Manual protobuf decoding
+    let mut data = data.0;
+    // Parse (optional) data
+    let data = parse_protobuf_bytes(&mut data)?;
+
+    Ok(MsgExecuteContractResponse { data })
+}
+
+#[derive(Error, Debug, PartialEq)]
+pub enum ParseReplyError {
+    #[error("Failure response from sub-message: {0}")]
+    SubMsgFailure(String),
+
+    #[error("Invalid reply from sub-message: {0}")]
+    ParseFailure(String),
+
+    #[error("Error occurred while converting from UTF-8")]
+    FromUtf8(#[from] std::string::FromUtf8Error),
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use cosmwasm_std::{ContractResult, SubMsgExecutionResponse};
+    use prost::Message;
+
+    #[derive(Clone, PartialEq, Message)]
+    pub struct MsgInstantiateContractResponse {
+        #[prost(string, tag = "1")]
+        pub contract_address: ::prost::alloc::string::String,
+        #[prost(bytes, tag = "2")]
+        pub data: ::prost::alloc::vec::Vec<u8>,
+    }
+
+    #[derive(Clone, PartialEq, Message)]
+    pub struct MsgExecuteContractResponse {
+        #[prost(bytes, tag = "1")]
+        pub data: ::prost::alloc::vec::Vec<u8>,
+    }
+
+    #[test]
+    fn parse_reply_instantiate_data_works() {
+        let contract_addr: &str = "Contract #1";
+        for data in [vec![], vec![1u8, 2, 255, 7, 5]] {
+            let expected = if data.is_empty() {
```

Comment:
If you are doing a loop ("table-tests" in Go speak), I would define input and expected to in a `vec![struct testData]` and then just test it. Building the expected based on the responses is a bit odd to read.